text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,485 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can also declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>`, in addition to using `Self` by value or reference.
+
+The `Receiver` trait is simple and only requires specifying the `Target` type:
+
+```rust
+trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+The `Receiver` trait is already implemented for many standard library types:
+- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<SomeSmartPtr<Self>>` (and in fact, any type which implements `Deref`)
+- references: `&Self` and `&mut Self`
+- pointers: `*const Self` and `*mut Self`
+
+Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.
+
+All of the following self types are valid:
+
+```rust
+impl Foo {
+    fn by_value(self /* self: Self */);
+    fn by_ref(&self /* self: &Self */);
+    fn by_ref_mut(&mut self /* self: &mut Self */);
+    fn by_ptr(self: *const Self);
+    fn by_mut_ptr(self: *mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_custom_ptr(self: CustomPtr<Self>);
+}
+
+struct CustomPtr<T>(*const T);
+
+impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+}
+```
+
+## Recursive arbitrary receivers
+
+Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.) so as we remove the special-casing we need to retain this property.
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
+}
+```
+
+The Rust language doesn't provide a way for user code to use this recursive property in generics or iteration, so this trait is unlikely to be useful except to the compiler. Nevertheless, we don't intend to _prevent_ use of the `Receiver` trait by user code: since the same recursive property applies to `Deref` yet it's been occasionally useful to [introduce `Deref` bounds](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.new_unchecked).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core` libs changes
+
+The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.
+
+This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.
+
+```rust
+pub trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+A blanket implementation is provided for any type that implements `Deref`:
+
+```rust
+impl<P: ?Sized> Receiver for P
+where
+    P: Deref,
+{
+    type Target = <P as Deref>::Target;
+}
+```
```

Comment:
Does the new `Receiver` trait also work when the target is embedded directly in the struct, rather than being an actual smart pointer? E.g.:
```Rust
struct Wrapper<T> {
    extra_data: Foo,
    inner: T,
}

impl MyStruct {
    fn method1(self: Wrapper<Self>);
    fn method2(self: Arc<Wrapper<Self>>);
}
```

Reply:
Yep. Receiver doesn't care about the internals, you can implement it on whatever you want